### ISSUE TITLE
Auto mining is started while vBitcoin not synced yet

### ISSUE THREAD

**Issue by Leonardiae** [VeriBlock/vbk-ri-btc#371]
Auto mining is started while vBitcoin not synced yet

```
 > getdebuginfo
getdebuginfo
Running several checks, this may take a few moments...
{
  "information": [
    "SUCCESS - SPV connection: Connected",
    "SUCCESS - SPV synchronization status: Synchronized 0 blocks left (LocalHeight=996559 NetworkHeight=996559 InitialBlockDownload=false)",
    "SUCCESS - vBitcoin connection: Connected to http://localhost:18332",
    "SUCCESS - vBitcoin configured network: vBitcoin & APM are running on the same configured network (testnet)",
    "SUCCESS - vBitcoin synchronization status: Synchronized 0 blocks left (LocalHeight=7262 NetworkHeight=7262 InitialBlockDownload=false)",
    "SUCCESS - vBitcoin payoutAddress: 'tb1qdtdxzxrsusdw70ceu9f9w5mayuhp20q4rxltrz' is configured",
    "SUCCESS - vBitcoin configuration for autoMineRounds is valid",
    "SUCCESS - The tVBK PoP wallet contains sufficient funds, current balance: 9.99012000 tVBK"
  ]
}
2020-11-03 20:11:32 si-poll    INFO  AltchainPoPMiner          - Unable to connect to vBitcoin chain at http://localhost:18332, trying to reconnect...
2020-11-03 20:11:42 si-poll    INFO  AltchainPoPMiner          - The connected vbtc chain & APM are running on the same configured network (testnet)
2020-11-03 20:11:42 si-poll    INFO  SecurityInheritingMonitor - The connected vBitcoin chain is synchronized: 0 blocks left (LocalHeight=7125 NetworkHeight=7125 InitialBlockDownload=false)
2020-11-03 20:11:42 si-poll    INFO  AltchainPoPMiner          - vbtc chain is ready
2020-11-03 20:11:42 si-poll    INFO  SecurityInheritingMonitor - Starting continuous submission of VBK Context for vBitcoin
2020-11-03 20:11:42 si-poll    INFO  SecurityInheritingMonitor - Starting continuous submission of VTBs for vBitcoin
2020-11-03 20:11:43 si-poll    INFO  BitcoinFamilyChain        - Retrieving mining instruction at height 7127 from vBitcoin daemon at http://localhost:18332...
2020-11-03 20:11:43 si-poll    INFO  BitcoinFamilyChain        - Submitting PoP data to vBitcoin daemon at http://localhost:18332...
2020-11-03 20:11:44 peer-input INFO  PeerSocketHandler         - Disconnected from peer /94.130.147.194:7501.
2020-11-03 20:11:44 peer-input INFO  PeerSocketHandler         - Output stream thread shutting down
2020-11-03 20:11:48 si-poll    WARN  SecurityInheritingMonitor - Error while submitting VTBs to vBitcoin! Will try again later...: Socket timeout has been expired [url=http://localhost:18332/, socket_timeout=unknown] ms
2020-11-03 20:11:48 si-poll    WARN  SecurityInheritingMonitor - Error while submitting context to vBitcoin! Will try again later...: Socket timeout has been expired [url=http://localhost:18332/, socket_timeout=unknown] ms
```

**Comment by Leonardiae** [comment]
Listoperations shows that blocks were indeed mined and paid for:

```
 > listoperations
listoperations
vbtcf9f90384: vBitcoin (7129 -> 7179) | VBK Publications submitted | Confirm ATV
vbtc47e8c869: vBitcoin (7137 -> 7187) | VBK Publications submitted | Confirm ATV
vbtcc21f1e79: vBitcoin (7148 -> 7198) | VBK Publications submitted | Confirm ATV
vbtcdc1af3e3: vBitcoin (7161 -> 7211) | VBK Publications submitted | Confirm ATV
vbtc4c2b5039: vBitcoin (7173 -> 7223) | VBK Publications submitted | Confirm ATV
vbtc3b029f3a: vBitcoin (7176 -> 7226) | VBK Publications submitted | Confirm ATV
vbtc9bf3561b: vBitcoin (7182 -> 7232) | VBK Publications submitted | Confirm ATV
vbtc9a8b1041: vBitcoin (7196 -> 7246) | VBK Publications submitted | Confirm ATV
vbtc8415eb38: vBitcoin (7312 -> 7362) | VBK Publications submitted | Confirm ATV
vbtc68ef6286: vBitcoin (7313 -> 7363) | VBK Publications submitted | Confirm ATV

```

**Comment by Leonardiae** [comment]
To be precise this occured when the APM was connected to the syncing vBTC node and I sended tVBK to the APM. As soon as the APM was funded it started mining.

**Comment by Warchant** [comment]
https://track.veriblock.org/issue/ALT-258
No changes from APM needed. Will be fixed in vBTC.

**Comment by Warchant** [comment]
Fixed by #377

**Comment by Leonardiae** [comment]
@Warchant Thanks for improving the header sync speed! But isn't the issue here that the APM starts mining before vBTC is synced? i.e. shouldnt the APM check if vBTC is synced before allowing an auto mine to start?

Getdebuginfo in APM shows if vBTC is synced so i guess this should be possible:
```
{
  "information": [
    "SUCCESS - SPV connection: Connected",
    "SUCCESS - SPV synchronization status: Synchronized 0 blocks left (LocalHeight=1015683 NetworkHeight=1015683 InitialBlockDownload=false)",
    "SUCCESS - vBitcoin connection: Connected to http://localhost:18332",
    "SUCCESS - vBitcoin configured network: vBitcoin & APM are running on the same configured network (testnet)",
`    "SUCCESS - vBitcoin synchronization status: Synchronized 0 blocks left (LocalHeight=8504 NetworkHeight=8504 InitialBlockDownload=false)",`
    "SUCCESS - vBitcoin payoutAddress: 'tb1qdtdxzxrsusdw70ceu9f9w5mayuhp20q4rxltrz' is configured",
    "SUCCESS - vBitcoin configuration for autoMineRounds is valid",
    "SUCCESS - The tVBK PoP wallet contains sufficient funds, current balance: 76.23215896 tVBK"
  ],
```

**Comment by Warchant** [comment]
APM starts mining because it used `getblockchaininfo` RPC call and saw 
```
blocks: X
headers: X
initialblockdownload: false
```
Where `blocks == headers`.

In original Bitcoin clients can decide if node is **not** in sync if `initialblockdownload == true || blocks != headers`.

When POP has been added to vBTC, we were not able to do POP-based fork resolution on headers because they do not have POP data (it is stored in block body), therefore our `headers` was always equal to `blocks`, which caused clients to think that node is already synchronized.

PR #377 fixes this by adding new condition - we initially do POW-based fork resolution on headers (without block bodies), and then if best block is on different chain than best known header, we fallback to old algorithm (when headers == blocks). 

**TLDR**: APM starts mining during vBTC sync, because vBTC has no good way to determine if it is in sync, and what is the best network block. It can determine this information when it downloads ALL blocks from the network and executes POP Fork Resolution on all of them.

**Comment by Leonardiae** [comment]
@Warchant Now I understand, thanks for explaining!
